fontbe/src/metrics_and_limits.rs

```diff
@@ -31,57 +35,178 @@ struct GlyphLimits {
-        let left_side_bearing = bbox.x_min;
-        // aw - (lsb + xMax - xMin) ... but if lsb == xMin then just advance - xMax?
-        let right_side_bearing: i16 = match advance as i32 - bbox.x_max as i32 {
-            value if value < i16::MIN as i32 => i16::MIN,
-            value if value > i16::MAX as i32 => i16::MAX,
-            value => value as i16,
-        };
-        self.min_left_side_bearing = self
-            .min_left_side_bearing
-            .map(|v| min(v, left_side_bearing))
-            .or(Some(left_side_bearing));
-        self.min_right_side_bearing = self
-            .min_right_side_bearing
-            .map(|v| min(v, right_side_bearing))
-            .or(Some(right_side_bearing));
-        self.x_max_extent = self
-            .x_max_extent
-            .map(|v| max(v, bbox.x_max))
-            .or(Some(bbox.x_max));
-        self.advance_width_max = max(self.advance_width_max, advance);
-
         match glyph {
             Glyph::Simple(simple) => {
-                let points = simple.contours().iter().map(Contour::len).sum::<usize>() as u16;
-                self.max_points = max(self.max_points, points);
-                self.max_contours = max(self.max_contours, simple.contours().len() as u16);
+                let num_points = simple.contours().iter().map(Contour::len).sum::<usize>() as u16;
+                let num_contours = simple.contours().len() as u16;
+                self.max_points = max(self.max_points, num_points);
+                self.max_contours = max(self.max_contours, num_contours);
+                self.glyph_info.insert(
+                    id,
+                    GlyphInfo {
+                        gid: id,
+                        num_points,
+                        num_contours,
+                        ..Default::default()
+                    },
+                )
             }
             Glyph::Composite(composite) => {
                 let num_components = composite.components().len() as u16;
                 self.max_component_elements = max(self.max_component_elements, num_components);
+                let components = Some(composite.components().iter().map(|c| c.glyph).collect());
+                self.glyph_info.insert(
+                    id,
+                    GlyphInfo {
+                        gid: id,
+                        components,
+                        ..Default::default()
+                    },
+                )
             }
         };
     }
+
+    fn compute_component_limits(&self) -> ComponentLimits {
+        let mut depths = HashMap::<GlyphId, u16>::new();
+        let mut pending = self
+            .glyph_info
+            .iter()
+            .filter_map(|(gid, gi)| if gi.is_component() { Some(*gid) } else { None })
+            .collect::<Vec<_>>();
+        let mut components = Vec::with_capacity(8);
+        let mut max_points = 0;
+        let mut max_contours = 0;
+        while !pending.is_empty() {
```

Comment:
a recursive algorithm like fonttools does would be much simpler to implement, is that a no-no in Rust? why?

Reply:
Recursion is fine in Rust. I might argue the simplicity of recursive code to be subjective ;)

Reply:
I tidied up the loop, I believe it is now much clearer. Please reopen if you disagree.